Azure ACA Dynamic Sessions tool and interfaces

bbenz opened the pull request:
<!--
Thank you so much for your contribution!

Please fill in all the sections below.
Please open the PR as a draft initially. Once it is reviewed and approved, we will ask you to add documentation and examples.
Please note that PRs with breaking changes or without tests will be rejected.

Please note that PRs will be reviewed based on the priority of the issues they address.
We ask for your patience. We are doing our best to review your PR as quickly as possible.
Please refrain from pinging and asking when it will be reviewed. Thank you for understanding!
-->

## Issue
<!-- Please specify the ID of the issue this PR is addressing. For example: "Closes #1234" or "Fixes #1234" -->
Closes #

## Change
Created a tool and interfaces for accessing Azure Container Apps (ACA) Dynamic Sessions using langchain4j.
More info here: Serverless code interpreter sessions in Azure Container Apps
https://learn.microsoft.com/en-us/azure/container-apps/sessions-code-interpreter
Also simultaneously sending a PR with examples to langchain4j-examples.


## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [X] There are no breaking changes
- [X] I have added unit and/or integration tests for my change
- [ ] The tests cover both positive and negative cases
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [X] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [X] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)


## Checklist for adding new maven module
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,125 @@
+
+        <dependency>
+            <groupId>com.squareup.okhttp3</groupId>
+            <artifactId>okhttp</artifactId>
```
please avoid using okhttp, it is better to use `dev.langchain4j.http.client.HttpClient` abstraction so that we can switch to another one when needed (e.g., in Quarkus or Spring Boot starters)

bbenz replied:
Thanks!  Converted to dev.langchain4j.http.client.HttpClient in the latest commit.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,125 @@
+
+        <dependency>
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
```
Is this required? Please use jackson

bbenz replied:
Thanks!  Converted to Jackson in the latest commit.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/src/main/java/dev/langchain4j/azure/aca/dynamicsessions/SessionsREPLTool.java:
Would it make sense for implement `CodeExecutionEngine` interface so that it can be used interchangeably with other existing implementations?

bbenz replied:
Great suggestion!  Converted to CodeExecutionEngine in the latest commit.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/src/main/java/dev/langchain4j/azure/aca/dynamicsessions/SessionsREPLTool.java:
Was it intended to name it My...?

bbenz replied:
Good catch!  renamed Myxxx to xxxImpl for all instances.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,125 @@
+
+        <dependency>
+            <groupId>com.azure</groupId>
+            <artifactId>azure-ai-openai</artifactId>
```
Is this required?

bbenz replied:
Unnecessary items removed.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,125 @@
+
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-azure-open-ai</artifactId>
```
Is this required? Please check every dependency in this module, it seems that many are not required

bbenz replied:
Unnecessary items removed.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,125 @@
+
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-azure-open-ai</artifactId>
```
Is this required? Please check every dependency in this module, it seems that many are not required

bbenz replied:
Resolved the ones I could resolve - 

Some are legit - example - runtime dependencies like langchain4j-http-client-jdk: These dependencies are correctly marked with runtime scope because they're needed at runtime but not at compile time. Maven's dependency analyzer sometimes flags these as "unused" because it only checks for compile-time references.

For test dependencies (junit-jupiter, mockito, assertj): These are used in the test code but flagged because the analyzer is looking at the main code.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/src/main/java/dev/langchain4j/azure/aca/dynamicsessions/SessionsREPLTool.java:
there is `httpClient` and `client`, why?

bbenz replied:
Again, good catch!  De-duped and eliminated the redundant names.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/src/main/java/dev/langchain4j/azure/aca/dynamicsessions/SessionsREPLTool.java:
`name` and `description` are not used

bbenz replied:
Thanks!  Iterative Leftovers.  Removed.

dliubarskyi reviewed (commented):
Hi @bbenz, thank you

dliubarskyi commented:
@bbenz there are also no tests

bbenz commented:
> Hi @bbenz, thank you

Thanks for taking the time to review and the feedback @dliubarskyi !  I'll make updates and respond to each item individually.

bbenz commented:
Thanks again for the feedback!  I think I've resolved all of the items above and added tests.  
Also added an example file with documentation for getting started: https://github.com/langchain4j/langchain4j-examples/pull/142
We will be adding docs and videos at Microsoft once merged.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+    <!-- Project Coordinates -->
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j-azure-aca-dynamic-sessions</artifactId>
+    <version>1.0.0-beta1</version>
```
```suggestion
```

bbenz replied:
Will keep - Had issue between betas, for example, using the parent 1.2.0-beta8-SNAPSHOT for the parent but some modules needed 1.2.0-SNAPSHOT.  Using the project version for those modules instead of the parent version.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+
+    <!-- Project Coordinates -->
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j-azure-aca-dynamic-sessions</artifactId>
```
please add this artifact to the root pom and langchain4j-bom

bbenz replied:
Done!

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+    <parent>
+        <groupId>dev.langchain4j</groupId>
+        <artifactId>langchain4j-parent</artifactId>
+        <version>1.0.0-beta1</version>
```
```suggestion
        <version>1.1.0-beta7-SNAPSHOT</version>
```

bbenz replied:
Upgraded the parent to 1.2.0-beta8-SNAPSHOT

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+    <packaging>jar</packaging>
+    <name>LangChain4j :: Integration :: Azure ACA Dynamic Sessions</name>
+
+    <licenses>
```
nit: redundant, will inherit from parent

bbenz replied:
Removed!

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+
+    <properties>
+        <!-- TODO: remove enforcer.skipRules -->
+        <enforcer.skipRules>dependencyConvergence</enforcer.skipRules>
```
```suggestion
```

bbenz replied:
Removed!

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-core</artifactId>
+            <version>${project.version}</version>
```
```suggestion
            <version>1.1.0-SNAPSHOT</version>
```

bbenz replied:
Will keep - using the parent 1.2.0-beta8-SNAPSHOT for the parent but some modules needed 1.2.0-SNAPSHOT. Using the project version for those modules instead of the parent version.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+        <!-- Dependencies for CodeExecutionEngine interface - langchain4j dependency is transitive through langchain4j-core -->
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j</artifactId>
```
why is this dependency required? `CodeExecutionEngine` is in the `langchain4j-core`

bbenz replied:
Removed!

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-http-client</artifactId>
+            <version>${project.version}</version>
```
```suggestion
            <version>1.1.0-SNAPSHOT</version>
```

bbenz replied:
Will keep - using 1.2.0-beta8-SNAPSHOT for the parent but some modules needed 1.2.0-SNAPSHOT. Using the project version for those modules instead of the parent version.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-http-client-jdk</artifactId>
+            <version>${project.version}</version>
```
```suggestion
            <version>1.1.0-SNAPSHOT</version>
```

bbenz replied:
Will keep - using 1.2.0-beta8-SNAPSHOT for the parent but some modules needed 1.2.0-SNAPSHOT. Using the project version for those modules instead of the parent version.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>2.18.2</version>
```
```suggestion
```
it will inherit it from parent

bbenz replied:
Updated to 2.19.0

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+        <dependency>
+            <groupId>com.azure</groupId>
+            <artifactId>azure-core</artifactId>
+            <version>1.54.1</version>
```
```suggestion
```
it will inherit it from parent (azure-sdk-bom)

bbenz replied:
Removed!

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+        <dependency>
+            <groupId>com.azure</groupId>
+            <artifactId>azure-identity</artifactId>
+            <optional>true</optional>
```
Seems like this is not an optional dependency, `SessionsREPLTool` explicitly imports it.
But perhaps it should be optional? cc @jdubois

bbenz replied:
Good catch!  Removed the optional element.

bbenz replied:
I spoke too soon :)  
I've marked it as optional in the POM to avoid dependency convergence issues with Netty libraries. Without the optional tag, multiple versions of Netty components are pulled in via different dependency paths. I've added documentation to clarify that azure-identity is a dependency.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+    </parent>
+
+    <!-- Project Coordinates -->
+    <groupId>dev.langchain4j</groupId>
```
```suggestion
```
redundant, will inherit from parent

bbenz replied:
Removed!

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/pom.xml:
```diff
@@ -0,0 +1,118 @@
+        <!-- Required for reactor.core.publisher.Mono used in getToken -->
+        <dependency>
+            <groupId>io.projectreactor</groupId>
+            <artifactId>reactor-core</artifactId>
```
this dependency seems to be only required for the `test` scope

bbenz replied:
It's needed for more: Added  <scope>test</scope> and:

ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.14.0:compile (java-compile) on project langchain4j-azure-aca-dynamic-sessions: Compilation failure
[ERROR] /mnt/c/githublocal/langchain4j/langchain4j-azure-aca-dynamic-sessions/src/main/java/dev/langchain4j/azure/aca/dynamicsessions/SessionsREPLTool.java:[247,35] cannot access reactor.core.publisher.Mono
[ERROR]   class file for reactor.core.publisher.Mono not found

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/src/main/java/dev/langchain4j/azure/aca/dynamicsessions/SessionsREPLTool.java:
nit: please rename classes into `Default*` instead of `*Impl`

bbenz replied:
Good call - Done!

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/src/main/java/dev/langchain4j/azure/aca/dynamicsessions/SessionsREPLTool.java:
This string needs to be updated. Do you plan to update it for every version change?

bbenz replied:
Good call - simplified the USER_AGENT

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/src/main/java/dev/langchain4j/azure/aca/dynamicsessions/SessionsREPLTool.java:
not used

bbenz replied:
Good catch!  Removed.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/src/main/java/dev/langchain4j/azure/aca/dynamicsessions/SessionsREPLTool.java:
please remove all `System.out...`

bbenz replied:
Done!

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/src/main/java/dev/langchain4j/azure/aca/dynamicsessions/SessionsREPLTool.java:
It is a bit wasteful to create new `ObjectMapper` every time (in 4 different places here). Please use a constant

bbenz replied:
Good call - I originally coded each separately but there's no need for the redundancy when they're combined.

dliubarskyi commented on langchain4j-azure-aca-dynamic-sessions/src/test/java/dev/langchain4j/azure/aca/dynamicsessions/SessionsREPLToolTest.java:
```diff
@@ -0,0 +1,169 @@
+     * Custom SessionsREPLTool for testing that uses mocked dependencies
+     */
+    private static class TestSessionsREPLTool extends SessionsREPLTool {
+        private final HttpClient mockHttpClient;
```
not used

dliubarskyi reviewed (commented):
@bbenz thank you!

bbenz commented:
This PR was automatically closed while I was trying to resolve fork/clone conflicts.  I've updated the code based on the most recent feedback (https://github.com/langchain4j/langchain4j/pull/2521#pullrequestreview-2904325448) and created a new PR - 
https://github.com/langchain4j/langchain4j/pull/3266
